Java/core/src/main/java/com/amazon/randomcutforest/store/IntervalManager.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.store;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.Arrays;
+
+/**
+ * This class defines common functionality for Store classes, including
+ * maintaining the stack of free pointers.
+ */
+public class IntervalManager {
```

Comment:
I assume you didn't want to create confusion with the existing Index Manager, but the methods in this class are `takeIndex` and `releaseIndex`, it's unclear what this has to do with intervals.

Reply:
Yes (re not wanting to change IndexManager). This class should replace the IndexManager. This class uses intervals to manage the free indices (which seem to be consumed and released in bursts of intervals in PointStore). The current IndexManager has the implicit way of compressing the stack etc. This new class simplifies all of that (keeping the intended implicit compression). It also eliminates the BitSet etc., and is easier to change it's size dynamically.